pkg/web_app/lib/src/search.dart

```diff
@@ -4,13 +4,50 @@
 
 import 'dart:html';
 
+import 'package:web_app/src/page_data.dart';
+
 void setupSearch() {
+  _setBackToReferrer();
   _setEventForSearchInput();
   _setEventForFiltersToggle();
   _setEventForSortControl();
   _setEventForCheckboxChanges();
 }
 
+void _setBackToReferrer() {
+  final bannerContainerElem = document.getElementById('banner-container');
+  if (bannerContainerElem == null) return;
+
+  // abort if not on a package page
+  if (!pageData.isPackagePage) return;
+
+  // TODO: remove this after finalizing the new new design
+  // abort if on old design
+  if (document.body.classes.contains('non-experimental')) return;
+
+  // abort if there is no referrer
+  final ref = document.referrer;
+  if (ref == null || ref.isEmpty) return;
+
+  // abort if the scheme://hostname[:port]/ is different
+  // - location.protocol already contains the colon
+  // - location.host already contains the :port
+  final loc = window.location;
+  final prefix = '${loc.protocol}//${loc.host}/';
+  if (!ref.startsWith(prefix)) return;
+
+  // abort if the referrer was not search
+  if (!ref.contains('/packages')) return;
```

Comment:
Couldn't we check that there is a `q` and that it's one of the search urls we think it's supposed to be? Not a dartdoc page or other things?

It seems we should prefer to be overly cautious here.  I'd rather we don't support a particular kind of special search page, than that we have a security vulnerability.

Reply:
I've updated this part a bit with proper `Uri` parsing, however, I wouldn't include the `q` parameter: it could be that the user is filtering for platform, it is on page 10, and it would be still helpful to go back to that page.